src/ripple/app/tx/impl/DID.cpp

```diff
@@ -0,0 +1,259 @@
+    }
+
+    return preflight2(ctx);
+}
+
+TER
+DIDSet::preclaim(PreclaimContext const& ctx)
+{
+    AccountID const account = ctx.tx[sfAccount];
+    Keylet const didKeylet = keylet::did(account);
+
+    if (!ctx.view.read(didKeylet))
+    {
+        if (!ctx.tx.isFieldPresent(sfURI) &&
+            !ctx.tx.isFieldPresent(sfDIDDocument))
+            return tecEMPTY_DID;
+    }
+
+    return tesSUCCESS;
+}
+
+TER
+addSLE(
+    ApplyContext& ctx,
+    std::shared_ptr<SLE> const& sle,
+    AccountID const& owner)
+{
+    auto const sleAccount = ctx.view().peek(keylet::account(owner));
+    if (!sleAccount)
+        return tefINTERNAL;
+
+    // Check reserve availability for new object creation
+    {
+        auto const balance = STAmount((*sleAccount)[sfBalance]).xrp();
+        auto const reserve =
+            ctx.view().fees().accountReserve((*sleAccount)[sfOwnerCount] + 1);
+
+        if (balance < reserve)
+            return tecINSUFFICIENT_RESERVE;
+    }
+
+    // Add ledger object to ledger
+    ctx.view().insert(sle);
+
+    // Add ledger object to owner's page
+    {
+        auto page = ctx.view().dirInsert(
+            keylet::ownerDir(owner), sle->key(), describeOwnerDir(owner));
+        if (!page)
+            return tecDIR_FULL;
+        (*sle)[sfOwnerNode] = *page;
+    }
+    adjustOwnerCount(ctx.view(), sleAccount, 1, ctx.journal);
+    ctx.view().update(sleAccount);
+
+    return tesSUCCESS;
+}
+
+TER
+DIDSet::doApply()
+{
+    // Edit ledger object if it already exists
+    Keylet const didKeylet = keylet::did(account_);
+    if (auto const sleDID = ctx_.view().peek(didKeylet))
+    {
+        if (auto const uri = ctx_.tx[~sfURI])
+        {
+            if (uri->empty())
+            {
+                sleDID->makeFieldAbsent(sfURI);
+            }
+            else
+            {
+                (*sleDID)[sfURI] = *uri;
+            }
+        }
+        if (auto const document = ctx_.tx[~sfDIDDocument])
+        {
+            if (document->empty())
+            {
+                sleDID->makeFieldAbsent(sfDIDDocument);
+            }
+            else
+            {
+                (*sleDID)[sfDIDDocument] = *document;
+            }
+        }
+
+        if (auto const attestation = ctx_.tx[~sfAttestation])
+        {
+            if (attestation->empty())
+            {
+                sleDID->makeFieldAbsent(sfAttestation);
+            }
+            else
+            {
+                (*sleDID)[sfAttestation] = *attestation;
+            }
+        }
+        if (!sleDID->isFieldPresent(sfURI) &&
+            !sleDID->isFieldPresent(sfDIDDocument))
+        {
+            return tecEMPTY_DID;
+        }
+        ctx_.view().update(sleDID);
+        return tesSUCCESS;
+    }
+
+    // Create new ledger object otherwise
+    auto const sleDID = std::make_shared<SLE>(didKeylet);
+    (*sleDID)[sfAccount] = account_;
+    if (auto const uri = ctx_.tx[~sfURI]; uri.has_value() && !uri->empty())
```

Comment:
Consider adding lambda:
>       auto set = [&](auto const& sField) {
>        if (auto const field = ctx_.tx[~sField];
>           field.has_value() && !field->empty())
>            (*sleDID)[sField] = field.value();
>      };
>      set(sfURI);
>      set(sfDIDDocument);
>      set(sfAttestation);

Reply:
Done in https://github.com/XRPLF/rippled/pull/4636/commits/04066ae42e9c8d7de59742d984b4202e4d705d54